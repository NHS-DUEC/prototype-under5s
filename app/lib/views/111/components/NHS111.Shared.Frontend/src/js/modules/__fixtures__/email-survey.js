export const formHtml = `<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit type="submit">Send</button>
    <p data-feedback-nojs-message>(Opens in a new tab or window)</p>
</div>
</form>`;
