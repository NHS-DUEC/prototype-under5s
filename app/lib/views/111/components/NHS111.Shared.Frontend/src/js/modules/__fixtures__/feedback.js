export const formHtml = `<form
data-test-id="feedback-content"
action="/Feedback/SubmitFeedback"
method="post"
target="_blank"
data-feedback-form
autocomplete="off"
>
<input
  type="hidden"
  name="UserId"
  value="c0604133-7150-442d-afbe-957f19902546"
/>

<input type="hidden" name="PageData.Page" value="ModuleZero" />
<input type="hidden" name="PageData.TxNumber" value="" />
<input type="hidden" name="PageData.QuestionId" value="" />
<input type="hidden" name="PageData.StartingPathwayNo" value="" />
<input type="hidden" name="PageData.StartingPathwayTitle" value="" />
<input type="hidden" name="PageData.Gender" value="Indeterminate" />
<input type="hidden" name="PageData.Age" value="" />
<input type="hidden" name="PageData.PathwayNo" value="" />
<input type="hidden" name="PageData.PathwayTitle" value="" />
<input type="hidden" name="PageData.DxCode" value="" />
<input type="hidden" name="PageData.SearchString" value="" />
<input type="hidden" name="PageData.Campaign" value="" />
<input type="hidden" name="PageData.Source" value="" />
<input type="hidden" name="PageData.StartUrl" value="default" />

<div class="nhsuk-form-group" data-feedback-form-section>
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input
          class="nhsuk-radios__input"
          id="feedback-choice-yes"
          name="feedbackChoice"
          type="radio"
          value="yes"
        />
        <label
          class="nhsuk-label nhsuk-radios__label"
          for="feedback-choice-yes"
          >Yes</label
        >
      </div>

      <div class="nhsuk-radios__item">
        <input
          class="nhsuk-radios__input"
          id="feedback-choice-no"
          name="feedbackChoice"
          type="radio"
          value="no"
        />
        <label
          class="nhsuk-label nhsuk-radios__label"
          for="feedback-choice-no"
          >No</label
        >
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section>
  <fieldset class="nhsuk-fieldset">
    <legend
      class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0"
      data-test-id="feedback-explanation-1"
    >
      Tell us how to make this page easier to understand
    </legend>

    <br />

    <div
      class="nhsuk-character-count"
      data-module="nhsuk-character-count"
      data-maxlength="1200"
    >
      <div class="nhsuk-form-group">
        <div
          class="app-callout app-callout--info"
          id="feedback-contact-warning"
        >
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label>
        <textarea
          class="nhsuk-textarea nhsuk-js-character-count"
          id="feedback-text"
          name="Text"
          rows="5"
          data-test-id="feedback-textarea"
          aria-describedby="feedback-text-info"
          data-autosize
        ></textarea>
      </div>

      <div
        class="nhsuk-hint nhsuk-character-count__message"
        id="feedback-text-info"
      >
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button
    data-test-id="send-feedback-button"
    data-event-trigger="click"
    data-event-value="Send feedback"
    class="nhsuk-button nhsuk-u-margin-bottom-2"
    type="submit"
  >
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message>
    (Opens in a new tab or window)
  </p>
</div>
</form>`;
