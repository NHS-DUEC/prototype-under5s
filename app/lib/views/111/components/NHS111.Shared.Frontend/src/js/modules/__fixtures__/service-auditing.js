export const serviceFixture = `
  <div class="service-details-wrapper" data-visible-service-id="1483092428"></div>
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Id" value="1483092428">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ServiceType.Id" value="46">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Name" value="Urgent Care Centre - Chorley" />
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Distance" value="12.6">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].PostCode" value="PR7 1PP">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].BookingStandard" value="CareConnect">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ActionAndReferral" value="GoTo-Referral">
`;

export const dosDataFixture = {
  "DosCheckCapacitySummaryResult.Success.Services[0].ActionAndReferral":
    "GoTo-Referral",
  "DosCheckCapacitySummaryResult.Success.Services[0].BookingStandard":
    "CareConnect",
  "DosCheckCapacitySummaryResult.Success.Services[0].Distance": "12.6",
  "DosCheckCapacitySummaryResult.Success.Services[0].Name":
    "Urgent Care Centre - Chorley",
  "DosCheckCapacitySummaryResult.Success.Services[0].PostCode": "PR7 1PP",
  "DosCheckCapacitySummaryResult.Success.Services[0].ServiceType.Id": "46",
  "DosCheckCapacitySummaryResult.Success.Services[0].Id": "1483092428",
};

export const serviceFixtureSingle = `
  <details class="service-listing" data-visible-service-id="111923" open>
    <div data-service-action="bookAndGo"></div>
    <div data-service-action="bookSlot" hidden></div>
    <div data-service-action="directions" hidden></div>
  </details>
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Id" value="111923">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ServiceType.Id" value="105">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Name" value="Emergency Department Royal Blackburn Hospital - Blackburn" />
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Distance" value="0.6">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].PostCode" value="BB2 3HH">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].BookingStandard" value="CareConnect">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ActionAndReferral" value="GoTo">
`;

export const CPCSFixture = `
  <div data-visible-service-id="1483092428">
    <div data-services-offered="Multiple CPCS offered">
    </div>
  </div>
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Id" value="1483092428">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ServiceType.Id" value="46">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Name" value="Urgent Care Centre - Chorley" />
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Distance" value="12.6">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].PostCode" value="PR7 1PP">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].BookingStandard" value="CareConnect">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ActionAndReferral" value="GoTo-Referral">
`;

export const validationFixture = `
  <div class="form-group measure" data-visible-service-id="1483092428" data-validation-service>
  <div>
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Id" value="1483092428">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ServiceType.Id" value="46">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Name" value="Urgent Care Centre - Chorley" />
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].Distance" value="12.6">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].PostCode" value="PR7 1PP">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].BookingStandard" value="CareConnect">
  <input type="hidden" name="DosCheckCapacitySummaryResult.Success.Services[0].ActionAndReferral" value="Callback">
`;
