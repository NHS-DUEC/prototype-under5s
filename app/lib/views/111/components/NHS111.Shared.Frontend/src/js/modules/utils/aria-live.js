export function ariaAssertiveMessage(message) {
  const ariaAssertiveRegion = document.getElementById("aria-assertive");
  const el = document.createElement("p");
  el.textContent = message;
  ariaAssertiveRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}

export function ariaPoliteMessage(message) {
  const ariaPoliteRegion = document.getElementById("aria-polite");
  const el = document.createElement("p");
  el.textContent = message;
  ariaPoliteRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}
