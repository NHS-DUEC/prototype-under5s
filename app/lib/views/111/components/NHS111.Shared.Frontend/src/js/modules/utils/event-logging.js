import { getPartPostCode } from "./postcode-utils";

export function getHiddenFieldValue(name, nameEndsWith = false) {
  const fields = nameEndsWith
    ? document.querySelectorAll(`[name$="${name}"]`)
    : document.getElementsByName(name);

  if (fields.length > 0) {
    return fields[0].value;
  }

  return "";
}

export function getGender() {
  const gender = getHiddenFieldValue("Gender", true);

  if (["male", "female"].includes(gender.toLowerCase())) return gender;

  return "";
}

export function getDxCode() {
  // find the DxCode by searching for any input with value starting Dx
  const dxCode = document.querySelector('input[value^="Dx"]');

  if (dxCode) {
    return dxCode.value;
  }

  return "";
}

export function createAuditEntry(
  eventKey,
  eventValue,
  eventData = "",
  answerOrder = "",
  answerTitle = ""
) {
  return {
    journeyId: getHiddenFieldValue("JourneyId", true),
    campaign: getHiddenFieldValue("Campaign"),
    dxCode: getDxCode(),
    postCodePart: getPartPostCode(getHiddenFieldValue("CurrentPostcode", true)),
    pathwayId: getHiddenFieldValue("PathwayId", true),
    pathwayTitle: getHiddenFieldValue("PathwayTitle", true),
    campaignSource: getHiddenFieldValue("Source"),
    QuestionNo: getHiddenFieldValue("QuestionNo"),
    QuestionTitle: getHiddenFieldValue("Title"),
    age: getHiddenFieldValue("Age", true),
    gender: getGender(),
    eventKey,
    eventValue,
    eventData,
    page: window.location.pathname,
    AnswerOrder: answerOrder,
    AnswerTitle: answerTitle,
  };
}

export function logEvent(
  eventKey,
  eventValue,
  eventData,
  answerOrder,
  answerTitle
) {
  const formData = new FormData();
  const auditEntry = createAuditEntry(
    eventKey,
    eventValue,
    eventData,
    answerOrder,
    answerTitle
  );

  Object.entries(auditEntry).forEach((item) => {
    const [key, value] = item;
    formData.append(key, value);
  });

  if (navigator.sendBeacon) {
    navigator.sendBeacon("/Auditing/Log", formData);
  }
}

export function logClick(eventValue, answerOrder, answerTitle) {
  logEvent("Clicked", eventValue, "", answerOrder, answerTitle);
}
