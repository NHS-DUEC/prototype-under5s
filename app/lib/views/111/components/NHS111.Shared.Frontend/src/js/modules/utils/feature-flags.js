const featureFlags = JSON.parse(
  document.querySelector("#feature-flags").innerHTML
);

export function featureIsEnabled(featureName) {
  return featureFlags[featureName];
}
