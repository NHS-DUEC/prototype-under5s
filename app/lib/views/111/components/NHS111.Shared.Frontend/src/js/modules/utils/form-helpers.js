export function hideNoJsMessage(form) {
  Array.from(form.querySelectorAll("[data-feedback-nojs-message]")).forEach(
    (item) => item.remove()
  );
}

export function clearErrors(form) {
  const errors = form.querySelectorAll("[data-feedback-error]");
  const formGroups = form.querySelectorAll(".nhsuk-form-group");
  const elementsWithOriginalAria = form.querySelectorAll(
    "[data-original-aria-describedby]"
  );

  Array.from(errors).forEach((item) => item && item.remove());
  Array.from(formGroups).forEach(
    (item) => item && item.classList.remove("nhsuk-form-group--error")
  );
  Array.from(elementsWithOriginalAria).forEach((el) =>
    el.setAttribute("aria-describedBy", el.dataset.originalAriaDescribedby)
  );
}

export function hideAllFormSections(form, formSections) {
  clearErrors(form);

  Array.from(formSections).forEach((element) => {
    element.hidden = true;
  });
}

export function trackOriginalAria(form) {
  // Keep track of any original aria-describedBy attributes so we can restore them when clearing errors
  Array.from(form.querySelectorAll("fieldset, input, textarea")).forEach(
    (el) => {
      if (el.getAttribute("aria-describedBy") !== null) {
        el.setAttribute(
          "data-original-aria-describedby",
          el.getAttribute("aria-describedBy")
        );
      }
    }
  );
}

export function formError(
  form,
  ErrorSummary,
  fieldId,
  errorMessage,
  errorLink = ""
) {
  const template = document.createElement("template");

  clearErrors(form);

  template.innerHTML = `
    <div class="nhsuk-error-summary" data-feedback-error data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#${fieldId}">${errorLink || errorMessage}</a>
          </li>
        </ul>
      </div>
    </div>`;

  const previousErrorSummary = form.querySelector(".nhsuk-error-summary");
  if (previousErrorSummary) {
    previousErrorSummary.remove();
  }

  const errorSummary = template.content;

  form.prepend(errorSummary);
  ErrorSummary();

  const field = form.querySelector(`#${fieldId}`);
  const formGroup = field.closest(".nhsuk-form-group");

  const errorMessageTemplate = document.createElement("template");
  errorMessageTemplate.innerHTML = `<span data-feedback-error data-test-id="feedback-error-message" class="nhsuk-error-message" id="${fieldId}-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> ${errorMessage}
    </span>`;

  if (field.getAttribute("type") === "radio") {
    formGroup.querySelector("legend").after(errorMessageTemplate.content);
    const fieldset = formGroup.querySelector("fieldset");
    fieldset.setAttribute(
      "aria-describedby",
      `${
        fieldset.dataset.originalAriaDescribedby
          ? fieldset.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  } else {
    formGroup.querySelector("label").after(errorMessageTemplate.content);
    field.setAttribute(
      "aria-describedby",
      `${
        field.dataset.originalAriaDescribedby
          ? field.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  }

  formGroup.classList.add("nhsuk-form-group--error");
}
