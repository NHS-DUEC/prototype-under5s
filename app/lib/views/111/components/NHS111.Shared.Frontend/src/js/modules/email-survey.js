import * as formHelpers from "./utils/form-helpers";

export class EmailSurveyForm {
  constructor(form, ErrorSummary) {
    this.form = form;
    this.formSections = this.form.querySelectorAll(
      "[data-survey-form-section]"
    );
    this.button = this.form.querySelector("[data-email-survey-submit]");
    this.cancelButton = this.form.querySelector("[data-email-survey-cancel]");
    this.ErrorSummary = ErrorSummary;
  }

  init() {
    formHelpers.hideNoJsMessage(this.form);
    formHelpers.trackOriginalAria(this.form);
    this.form.addEventListener("submit", this.formSubmitHandler());

    if (this.cancelButton) {
      this.cancelButton.addEventListener("click", this.cancelHandler());
    }
  }

  formSubmitHandler() {
    return (e) => {
      e.preventDefault();
      this.submitEmail();
    };
  }

  setResponseHeaderAndMessage(responseHeader, responseMessage) {
    formHelpers.hideAllFormSections(this.form, this.formSections);

    const header = document.createElement("h3");
    header.innerHTML = responseHeader;
    this.formSections[0].after(header);

    const message = document.createElement("p");
    message.tabIndex = -1;
    message.innerHTML = responseMessage;
    message.classList.add("nhsuk-u-margin-bottom-0", "app-u-no-focus-outline");
    message.setAttribute("data-test-id", "survey-response");
    header.after(message);

    message.focus();
  }

  cancelHandler() {
    return (e) => {
      e.preventDefault();
      formHelpers.clearErrors(this.form);
      this.setResponseHeaderAndMessage(
        "Thanks for your feedback",
        "We will use it to improve the services we recommend."
      );
    };
  }

  submitEmail() {
    this.button.disabled = true;

    formHelpers.clearErrors(this.form);

    const formData = new FormData(this.form);

    return fetch(this.form.action, {
      method: this.form.method,
      body: formData,
      headers: {
        Accept: "application/json",
      },
    })
      .then((response) => response.json())
      .then((data) => {
        if (!data.IsValid) {
          this.button.disabled = false;

          formHelpers.formError(
            this.form,
            this.ErrorSummary,
            "email",
            data.Message
          );
        } else {
          this.setResponseHeaderAndMessage(data.Header, data.Message);
        }
      });
  }
}

export function initSurvey(ErrorSummary) {
  if (typeof window.fetch === "undefined") {
    return;
  }

  const form = document.querySelector("[data-survey-form]");

  if (form) {
    const emailSurveyForm = new EmailSurveyForm(form, ErrorSummary);
    emailSurveyForm.init();
  }
}
