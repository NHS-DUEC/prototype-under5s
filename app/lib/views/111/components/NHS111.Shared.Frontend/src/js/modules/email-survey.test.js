import ErrorSummary from "nhsuk-frontend/packages/components/error-summary/error-summary";

import { formHtml } from "./__fixtures__/email-survey";
import { initSurvey, EmailSurveyForm } from "./email-survey";
import * as formHelpers from "./utils/form-helpers";

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(JSON.stringify("Thankyou message")),
  })
);

describe("email survey", () => {
  beforeEach(() => {
    global.fetch.mockClear();
  });

  describe("initSurvey method", () => {
    test("doesn't throw error when no form is present", () => {
      document.body.innerHTML = "";

      initSurvey();

      expect(initSurvey).not.toThrow();
    });

    test("initialises survey class when form is present", () => {
      document.body.innerHTML = formHtml;

      initSurvey();

      expect(document.body.innerHTML).toMatchSnapshot();
    });

    test("bails out if fetch is not present and leaves the form intact (in no-js mode)", () => {
      document.body.innerHTML = formHtml;

      const tempFetch = global.fetch;

      delete global.fetch;
      initSurvey();
      global.fetch = tempFetch;

      expect(document.body.innerHTML).toMatchSnapshot();
    });
  });

  describe("init method", () => {
    test("sets up pages correctly", () => {
      document.body.innerHTML = formHtml;
      const trackOriginalAriaMock = jest.spyOn(
        formHelpers,
        "trackOriginalAria"
      );
      const hideNoJsMessageMock = jest.spyOn(formHelpers, "hideNoJsMessage");

      const emailSurveyForm = new EmailSurveyForm(
        document.querySelector("form"),
        ErrorSummary
      );
      emailSurveyForm.init();

      expect(trackOriginalAriaMock).toHaveBeenCalled();
      expect(hideNoJsMessageMock).toHaveBeenCalled();

      expect(document.body.innerHTML).toMatchSnapshot();
    });
    test("sets up form submit handler correctly", () => {
      // Submit feedback method and calls to window.fetch etc tested later, this test just needs to check the submit handler is wired up correctly
      document.body.innerHTML = formHtml;

      let submitHandlerCalled = false;

      const submitHandler = function dummySubmitHandler(e) {
        e.preventDefault();
        submitHandlerCalled = true;
      };

      const formSubmitHandlerMock = jest
        .spyOn(EmailSurveyForm.prototype, "formSubmitHandler")
        .mockImplementation(() => {
          return submitHandler;
        });

      initSurvey(ErrorSummary);

      document.querySelector("form").submit();

      expect(submitHandlerCalled).toEqual(true);

      formSubmitHandlerMock.mockRestore();
    });
  });
  describe("submitEmail method", () => {
    test("when data from response is invalid displays error content", async () => {
      document.body.innerHTML = formHtml;

      const formErrorMock = jest.spyOn(formHelpers, "formError");

      const clearErrorsMock = jest.spyOn(formHelpers, "clearErrors");

      const emailSurveyForm = new EmailSurveyForm(
        document.querySelector("form"),
        ErrorSummary
      );
      const expectedFormData = {
        "data-1": "value-1",
        "data-2": "value-2",
      };

      const response = {
        IsValid: false,
        Message: "The email address provided is not valid",
      };
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve(JSON.stringify(response)),
        })
      );

      await emailSurveyForm.submitEmail();

      expect(document.body.innerHTML).toMatchSnapshot();

      expect(global.fetch).toHaveBeenCalledWith(
        "http://localhost/Survey/EmailSurveyCollection",
        expect.objectContaining({
          body: expect.any(FormData),
          headers: {
            Accept: "application/json",
          },
          method: "post",
        })
      );

      expect(formErrorMock).toHaveBeenCalled();
      expect(clearErrorsMock).toHaveBeenCalled();

      const formData = Array.from(
        global.fetch.mock.calls[0][1].body.entries()
      ).reduce((acc, f) => ({ ...acc, [f[0]]: f[1] }), {}); // reduce form data to an object

      expect(formData).toMatchObject(expectedFormData);
    });

    test("when data from response is valid displays success content", async () => {
      document.body.innerHTML = formHtml;

      const hideAllFormSectionsMock = jest.spyOn(
        formHelpers,
        "hideAllFormSections"
      );

      const clearErrorsMock = jest.spyOn(formHelpers, "clearErrors");

      const emailSurveyForm = new EmailSurveyForm(
        document.querySelector("form"),
        ErrorSummary
      );
      const expectedFormData = {
        "data-1": "value-1",
        "data-2": "value-2",
      };

      const response = {
        IsValid: true,
        Header: "Help Improve 111",
        Message: "We have successfully sent your survey",
      };

      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve(response),
        })
      );

      await emailSurveyForm.submitEmail();

      expect(document.body.innerHTML).toMatchSnapshot();

      expect(global.fetch).toHaveBeenCalledWith(
        "http://localhost/Survey/EmailSurveyCollection",
        expect.objectContaining({
          body: expect.any(FormData),
          headers: {
            Accept: "application/json",
          },
          method: "post",
        })
      );

      expect(hideAllFormSectionsMock).toHaveBeenCalled();
      expect(clearErrorsMock).toHaveBeenCalled();

      const formData = Array.from(
        global.fetch.mock.calls[0][1].body.entries()
      ).reduce((acc, f) => ({ ...acc, [f[0]]: f[1] }), {}); // reduce form data to an object

      expect(formData).toMatchObject(expectedFormData);
    });
  });
});
