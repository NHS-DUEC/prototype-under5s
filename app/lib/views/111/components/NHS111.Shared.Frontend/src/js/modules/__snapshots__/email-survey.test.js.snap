// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email survey init method sets up pages correctly 1`] = `
"<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel="" type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit="" type="submit">Send</button>
    
</div>
</form>"
`;

exports[`email survey initSurvey method bails out if fetch is not present and leaves the form intact (in no-js mode) 1`] = `
"<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel="" type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit="" type="submit">Send</button>
    <p data-feedback-nojs-message="">(Opens in a new tab or window)</p>
</div>
</form>"
`;

exports[`email survey initSurvey method initialises survey class when form is present 1`] = `
"<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel="" type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit="" type="submit">Send</button>
    
</div>
</form>"
`;

exports[`email survey submitEmail method when data from response is invalid displays error content 1`] = `
"<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#email">undefined</a>
          </li>
        </ul>
      </div>
    </div>
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group nhsuk-form-group--error">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="email-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> undefined
    </span>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email" aria-describedby=" email-error">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel="" type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit="" type="submit">Send</button>
    <p data-feedback-nojs-message="">(Opens in a new tab or window)</p>
</div>
</form>"
`;

exports[`email survey submitEmail method when data from response is valid displays success content 1`] = `
"<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="" hidden="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel="" type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit="" type="submit" disabled="">Send</button>
    <p data-feedback-nojs-message="">(Opens in a new tab or window)</p>
</div><h3>Help Improve 111</h3><p tabindex="-1" class="nhsuk-u-margin-bottom-0 app-u-no-focus-outline" data-test-id="survey-response">We have successfully sent your survey</p>
</form>"
`;
